src/main/java/org/tugraz/sysds/runtime/matrix/data/LibMatrixReorg.java

```diff
@@ -356,23 +383,73 @@ public static MatrixBlock sort(MatrixBlock in, MatrixBlock out, int[] by, boolea
 			values[i] = in.quickGetValue(i, by[0]-1);
 		}
 		
-		//sort index vector on extracted data (unstable)
-		SortUtils.sortByValue(0, rlen, values, vix);
-		
-		//sort by secondary columns if required (in-place)
-		if( by.length > 1 )
-			sortBySecondary(0, rlen, values, vix, in, by, 1);
-		
-		//flip order if descending requested (note that this needs to happen
-		//before we ensure stable outputs, hence we also flip values)
-		if(desc) {
-			sortReverseDense(vix);
-			sortReverseDense(values);
+		// step 4: split the data into number of blocks of PAR_NUMCELL_THRESHOLD_SORT (1024) elements.
+		if (k == 1 || rlen < PAR_NUMCELL_THRESHOLD_SORT){ // There is no parallel
+			//sort index vector on extracted data (unstable)
+			SortUtils.sortByValue(0, rlen, values, vix);
+			
+			//sort by secondary columns if required (in-place)
+			if( by.length > 1 )
+				sortBySecondary(0, rlen, values, vix, in, by, 1);
+			
+			//flip order if descending requested (note that this needs to happen
+			//before we ensure stable outputs, hence we also flip values)
+			if(desc) {
+				sortReverseDense(vix);
+				sortReverseDense(values);
+			}
+					
+			//final pass to ensure stable output
+			sortIndexesStable(0, rlen, values, vix, in, by, 1);
+			
+		} else {
+			try {
+				//Timing time_internal = new Timing(true);
+				ExecutorService pool = CommonThreadPool.get(k);
+				//compute actual transpose and check for errors
+				// sort smaller blocks.
+				ArrayList<SortTask> tasks = new ArrayList<>();
+				int blklen = 10; 
+				
+				//int blklen = PAR_NUMCELL_THRESHOLD_SORT; 
+				//int blklen = (int)(Math.ceil((double)rlen/k));
+				//blklen += (blklen%8 != 0)?8-blklen%8:0;
+				System.out.println(blklen);
```

Comment:
Maybe use the log facility if you want to keep that output. See for example how LibMatrixNative does it.

Reply:
- It is unintended should be removed.
- Which logging facility and where do I find an example?

Reply:
Ok, a better example than LibMatrixNative would be in InterProceduralAnalysis.java lines 81+82,100-106,184+185. 
You define a log object, a log level and an on/off switch and use it like "if(logLevelIsSet) \ printMessage".